lib/posticle/src/lib.rs

```diff
@@ -40,72 +56,135 @@ impl Entity {
 pub fn entities(text: &str) -> Vec<Entity> {
-    if text.is_empty() {
-        return Vec::new();
-    }
+    let pairs = Grammar::parse(Rule::post, text).unwrap_or_else(|e| panic!("{}", e));
+    let mut results = Vec::new();
 
-    let mut results = extract_urls(text);
-    results.extend(extract_hashtags(text, &results));
-    results.extend(extract_mentions(text, &results));
+    for pair in pairs {
+        match match pair.as_rule() {
+            Rule::hashtag => Entity::from_hashtag(pair),
+            Rule::mention => Entity::from_mention(pair),
+            Rule::url => Entity::from_url(pair),
+            _ => unreachable!(),
+        } {
+            Some(entity) => results.push(entity),
+            None => {},
+        }
+    }
 
-    results.sort_by(|a, b| a.range.cmp(&b.range));
     results
 }
 
-/// Given `text`, extract all [URL](EntityKind::Url) entities.
-pub fn extract_urls(text: &str) -> Vec<Entity> {
-    regexes::RE_URL
-        .find_iter(text)
-        .map(|mat| Entity {
-            kind:  EntityKind::Url,
-            range: (mat.start(), mat.end()),
-        }).collect()
+/// Check that a hashtag name (after the first #) is valid.
+pub fn validate_hashtag_name(name: &str) -> bool {
+    VALID_HASHTAG_NAME_RE.is_match(name)
 }
 
-/// Given `text` and some `existing` entities, extract all [Hashtag](EntityKind::Hashtag) entities
-/// which do not overlap with the `existing` ones.
-pub fn extract_hashtags(text: &str, existing: &[Entity]) -> Vec<Entity> {
-    regexes::RE_HASHTAG
-        .find_iter(text)
-        .map(|mat| Entity {
-            kind:  EntityKind::Hashtag,
-            range: (mat.start(), mat.end()),
-        }).filter(|en| {
-            existing
-                .iter()
-                .all(|existing_en| !en.overlaps_with(existing_en))
-        }).collect()
+/// Check that a mentioned username is valid.
+pub fn validate_mention_username(username: &str) -> bool {
+    VALID_MENTION_USERNAME_RE.is_match(username)
 }
 
-/// Given `text` and some `existing` entities, extract all [Mention](EntityKind::Mention) entities
-/// which do not overlap with the `existing` ones.
-pub fn extract_mentions(text: &str, existing: &[Entity]) -> Vec<Entity> {
-    regexes::RE_MENTION
-        .captures_iter(text)
-        .map(|capt| {
-            let whole = capt.get(0).unwrap();
-            let user = capt[1].to_string();
-            let domain = capt.get(2).map(|s| s.as_str().to_string());
-            Entity {
-                kind:  EntityKind::Mention(user, domain),
-                range: (whole.start(), whole.end()),
-            }
-        }).filter(|en| {
-            existing
-                .iter()
-                .all(|existing_en| !en.overlaps_with(existing_en))
-        }).collect()
+/// Check that a mentioned instance domain is valid.
+pub fn validate_mention_domain(domain: &str) -> bool {
+    validator::validate_url(format!("https://{}", domain))
+}
+
+pub fn validate_url(url: &str) -> bool {
```

Comment:
why is this here?

Reply:
Right, it doesn't need to be, and the other validators don't need to be public either.

Reply:
i'm more concerned with why you're writing a wrapper function that does nothing but call the other function. you can just import it and reexport if you really want to

Reply:
It's just an artefact from development, it originally just returned false, then used the `url` crate, now the `validator` crate.